extensions/pooling/src/pooling.ts

```diff
@@ -0,0 +1,314 @@
+   */
+  poolOptions?: Options;
+}
+
+/**
+ * Life cycle methods that a poolable resource can optionally implement so that
+ * they can be triggered by the pooling service
+ */
+export interface Poolable extends LifeCycleObserver {
+  /**
+   * To be called right after the resource is acquired from the pool. If it
+   * fails, the resource will be destroyed from the pool. The method should be
+   * used to set up the acquired resource.
+   */
+  acquire?(): ValueOrPromise<void>;
+  /**
+   * To be called right before the resource is released to the pool. If it
+   * fails, the resource will be destroyed from the pool. This method should be
+   * used to clean up the resource to be returned.
+   */
+  release?(): ValueOrPromise<void>;
+}
+
+/**
+ * Pooled resource instance
+ */
+export interface PooledValue<T> {
+  /**
+   * The resource pool
+   */
+  pool: Pool<T>;
+  /**
+   * Acquired value from the pool
+   */
+  value: T;
+  /**
+   * The function to release the acquired value back to the pool
+   */
+  release(): Promise<void>;
+}
+
+/**
+ * Acquire a resource from the pooling service or pool
+ * @param poolingService - Pooling service or pool
+ */
+export async function getPooledValue<T>(
+  poolingService: PoolingService<T> | Pool<T>,
+): Promise<PooledValue<T>> {
+  const value = await poolingService.acquire();
+  const pool =
+    poolingService instanceof PoolingService
+      ? poolingService.pool
+      : poolingService;
+  return {
+    pool,
+    value,
+    async release() {
+      return poolingService.release(value);
+    },
+  };
+}
+
+/**
+ * A singleton service to maintain a pool of resources. This pool service can
+ * be bound to different keys to represent multiple pools. Each binding is a
+ * singleton so that the state stays the same for injections into multiple
+ * instances for other artifacts.
+ *
+ * @remarks
+ *
+ * Some resources can be expensive to create/start. For example, a datasource
+ * has overhead to connect to the database. There will be performance penalty
+ * to use `TRANSIENT` scope and creates a new instance per request. But it is
+ * not feasible to be a singleton for some use cases, for example, each request
+ * may have different security contexts.
+ *
+ * The pool service observes life cycle events to start and stop.
+ */
+@bind({scope: BindingScope.SINGLETON})
+export class PoolingService<T> implements LifeCycleObserver {
+  private readonly factory: PoolFactory<T>;
+  /**
+   * The resource pool
+   */
+  readonly pool: Pool<T>;
+
+  constructor(
+    @inject.context() readonly context: Context,
+    @config() private options: PoolingServiceOptions<T>,
+  ) {
+    let factory = this.options.factory;
+    if (typeof factory === 'function') {
+      factory = factory(this.context);
+    }
+    this.factory = factory;
+    this.options = {...options, factory};
+    this.pool = createPool(factory, {
+      max: 8, // Default to 8 instances
+      ...this.options.poolOptions,
+      autostart: false,
```

Comment:
If `autostart` is to be configurable, it must be specified before `...this.options.poolOptions`.

Reply:
I intentionally disable `autostart` to fit into our life cycles. Added more comments in the code.